### ISSUE TITLE
outputs[0] is the zeroes tensor initialized by torch.zeroes()

### ISSUE THREAD

**Issue by wz337** [bentrevett/pytorch-seq2seq#45]
I have a question regarding the first element of the outputs (the prediction tensors). 

So in the class Seq2Seq(nn.Module), outputs gets initialized by calling "outputs = torch.zeros(max_len, batch_size, trg_vocab_size).to(self.device)." And in the for loop followed right after the initialization, we are starting from the 1st idx of outputs and put predicted tensor in each timestamp in outputs[t].

I am not sure whether I have explained it well, but shouldn't the outputs[0] always correspond to the start token <sos>? I am asking this because I have tried training the model and I am getting very weird output at outputs[0].

**Comment by bentrevett** [comment]
The `outputs` tensor is just a tensor to store all of the translations. Because the loop goes from 1 to `max_len - 1` and not from `0`, this means `outputs[0]` will stay full of zeros. This is fine, however, as before we compare it with the target sentence we cut off these zeros with `output[1:]`. 

Not sure if this answers your question, but this image may help:

![](https://i.imgur.com/JgRVBH3.jpg)

**Comment by wz337** [comment]
@bentrevett 

Thanks for the quick and detailed reply! And thanks for drawing this out. =) 

My concern is this won't be taking into the consideration when we are computing the loss, right? As from the tutorial, I am seeing we are ignoring "<pad>" by "criterion = nn.CrossEntropyLoss(ignore_index = PAD_IDX)" but not the init_token and eos_token. This might be a stupid question, but I am trying to understand the nuance. 

Thanks in advance!

**Comment by bentrevett** [comment]
You're correct. This is an example without padding. When there is padding in the sequence, which is very often the case, we have the padding tokens at the end of the sequence after the `<eos>` token. These are not cut off by the `trg[1:]` and are just left at the end. This is fine as, like you mentioned, we are ignoring the loss over the padding tokens.

![](https://i.imgur.com/M1qFqYa.jpg)

So, in the example above, we only calculate the loss across the three items in the sequence and the `<eos>` token, not across the padding tokens, which is exactly what we want.

One thing to note is that it only ignores the loss between items where the `trg` sequence has a padding token. In the example below, there model had incorrectly predicted a pad token where it should've predicted an `<eos>` token. The loss **is** calculated for that token because `trg` at that index is not a padding token. The loss is still not calculated over the final two tokens as both of those are padding tokens in `trg`. 

![](https://i.imgur.com/eNk0dE9.jpg)
